src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/StreamHelper.cs

```diff
@@ -276,26 +276,53 @@ internal static class StreamHelper
 
-        internal static void WriteToStream(Stream input, Stream output, PSCmdlet cmdlet, CancellationToken cancellationToken)
+        internal static void WriteToStream(Stream input, Stream output, WebRequestPSCmdlet cmdlet, CancellationToken cancellationToken)
         {
             if (cmdlet == null)
             {
                 throw new ArgumentNullException(nameof(cmdlet));
             }
 
-            Task copyTask = input.CopyToAsync(output, cancellationToken);
-
             ProgressRecord record = new(
                 ActivityId,
                 WebCmdletStrings.WriteRequestProgressActivity,
                 WebCmdletStrings.WriteRequestProgressStatus);
 
+            // If the client loses network connectivity for over 1 minute, the task returned from 
+            // 'CopyToAsync()' will never complete and therefore the loop below will be infinite.
+            // Short explaination:
+            // - CopyToAsync() uses ReadAsync() in loop until ReadAsync() returns 0, indicating EOF.
+            // - ReadAsync() read from network Socket.
+            // - If losing network connectivity for over 1 minute, the Socket will lose connect to target service
+            //   but OS never close the Socket (tested on Windows).
+            // - Then Socket never return anything, and thus ReadAsync() will be infinitely blocked on the Socket
+            //   and CopyToAsync() task will be never completed.
+            //
+            // Since cancelation token in CopyToAsync() applies to whole CopyToAsync()
+            // and it is not restarted for very internal ReadAsync()
+            // we have to use a workaround to cancel the ReadAsync() which is blocked on Socket longer Timeout.
+            // The workaround is to reset the cancelation timer
+            // while the length of the output file is constantly increasing.
+            long previousLength = 0;
+            var timeout = cmdlet.TimeoutSec == 0 ? Timeout.Infinite : cmdlet.TimeoutSec * 1000;
+            using var cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+            cts.CancelAfter(timeout);
+            Task copyTask = input.CopyToAsync(output, cts.Token);
+
             try
             {
-                while (!copyTask.Wait(1000, cancellationToken))
+                while (!copyTask.Wait(1000, cts.Token))
                 {
                     record.StatusDescription = StringUtil.Format(WebCmdletStrings.WriteRequestProgressStatus, output.Position);
                     cmdlet.WriteProgress(record);
+
+                    if (previousLength != output.Length)
+                    {
+                        // Reset cancelation timer while information continues to flow from network.
+                        // Cancelation timer applies only during no network connectivity.
+                        previousLength = output.Length;
+                        cts.CancelAfter(timeout);
+                    }
```

Comment:
This doesn't seem to be needed to me.
We should timeout just in the timespan specified by `-Timeout`, but what you are doing here is to refresh the timer after every 1 second.

Reply:
In the case you say the Timeout must be longer than the expected time to fully load the file. This immediately indicates a lot of problems. For example, the time to fully load a file is unpredictable even if there are no problems, because this time is determined by many factors (client, server, network). In general, even the size of the file is unknown. Thus, neither the script developer nor the user can set a reasonable timeout value - it will be very long for the script to accurately execute. But such a protective timeout is already possible in Task Scheduler or CIs - it can be an hour or two.
If you look from HttpClient side then before uploading it performs sending request and receiving response, these two operations are performed with the same timeout, but these are instantaneous operations - here you need timeout of a minute or two, but not an hour or two.
If you look at it from the service side, _any service closes the session when it loses contact with the client_ after a short time. With Kestrel, for example, this is 130 seconds by default. Usually it's faster (about a minute). _This means that it doesn't make sense for the client to wait an hour or two._ This also makes it clear that it makes sense to specify a timeout just for each request and response operation. That's exactly what this code does. We can explicitly specify in the documentation that if network connectivity is lost, cmdlet will crash after a given timeout.

Reply:
This is the doc for `-TimeoutSec`:
> Specifies how long the request can be pending before it times out. Enter a value in seconds. The default value, 0, specifies an indefinite time-out.

To an end user, it doesn't matter it's because of sending/receiving http request or downloading a file, when specifying `-TimtoutSec`, what the user wants is for `Invoke-WebRequest/RestMethod` to timeout if the operation cannot finish in X seconds. So, we need to make sure that's the case (maybe not accurately X seconds, but closely around that time).

Reply:
The Timeout did never work for download scenario in all PowerShell versions.. So we can not say "the user want" for the download operation.  In IO operations timeout is always protection from pending (which is just in the documentation)

Reply:
@JamesWTruher and @SeeminglyScience, can you please share your thoughts on this?
Mark the PR to be reviewed by cmdlet WG.